egauge/script/api_egauge.py

```diff
@@ -0,0 +1,148 @@
+#!./env/bin/python3
+"""
+This script makes egauge api requests using data retrieved from a database
+
+You can run this script using the form:
+python3 egauge_api.py <sensor id> <database url> <unit of time> <output file>
+
+Example command: python3 egauge_api.py 725 postgresql:///egauge m ./output.log
+"""
+
+from io import StringIO
+
+from sqlalchemy import create_engine
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.sql.expression import func #used for sql max() function
+
+import arrow
+import orm_egauge
+import os
+import pandas
+import requests
+import sqlalchemy
+import sys
+import traceback
+
+# used to initialize sqlalchemy ORM
+BASE = declarative_base()
+SCRIPT_NAME = os.path.basename(__file__)
+ERROR_LOG = 'error.log'
+
+def pull_egauge_data(sensor_id='31871', db_url='postgresql:///egauge', unit_of_time='m', output_file='./output.log'):
+    """
+    This function pulls egauge sensor data.
+
+    Data pulled will start at and include the time read from the timestamp log up to (but not including) the current time. The data (not including the headers) pulled from the sensors by this program will be appended to a csv file ('output.log' by default).
+    Any exceptions thrown during execution will be logged to a file.
+
+    Keyword arguments:
+        sensor_id: a string representing the id of the egauge sensor
+        unit_of_time: the unit of time we want the data returned as
+        output_file: the name of the file that egauge sensor data will be appended to
+    """
+
+    # cast sensor_id as int for consistency since id in database is int
+    sensor_id = int(sensor_id)
+    is_success = False
+    current_time = arrow.now().floor('minute')
+
+    try:  # should catch any database connection errors, timestamp errors, etc
+        # connect to the database
+        db = create_engine(db_url)
+        Session = sessionmaker(db)
+        session = Session()
+        BASE.metadata.create_all(db)
+
+        latest_retrieved_time = ''
+        last_inserted_time = ''
+        try:  # this try-except block should catch any egauge sensor api connection errors
+            # get the latest timestamp
+            rows = session.query(func.max(orm_egauge.DatabaseInsertionTimestamp.timestamp)).filter_by(is_success=True)
+
+            # if no timestamp is found, insert the current time and skip egauge api request, etc.
+            if not rows[0][0]:
```

Comment:
rows are vectors, I am confused why it would require rows[0][0] instead of rows[0]. Is this really a row being returned?

Reply:
From my understanding, session.query returns something similar to a vector of tuples. (I think it would be a list of tuples in the python terminology) https://stackoverflow.com/questions/48466959/query-for-list-of-attribute-instead-of-tuples-in-sqlalchemy

Even if only one column of data is being requested, any data in a tuple is still accessed like a vector/list.

Maybe I could rename 'rows' to 'timestamp_query_result' if that sounds better? 
Or add a comment noting that?

Reply:
Based on what we discussed on call, lets just call this `rows` variable as `latest_timestamp_sucessfuly_inserted`.